core/src/apps/ethereum/definitions.py

```diff
@@ -0,0 +1,150 @@
+from typing import TYPE_CHECKING
+
+from trezor.messages import EthereumNetworkInfo, EthereumTokenInfo
+from trezor.wire import DataError
+
+if TYPE_CHECKING:
+    from typing import TypeVar
+    from typing_extensions import Self
+
+    DefType = TypeVar("DefType", EthereumNetworkInfo, EthereumTokenInfo)
+
+
+def decode_definition(definition: bytes, expected_type: type[DefType]) -> DefType:
+    from trezor.crypto.cosi import verify as cosi_verify
+    from trezor.crypto.hashlib import sha256
+    from trezor.enums import EthereumDefinitionType
+    from trezor.protobuf import decode as protobuf_decode
+    from trezor.utils import BufferReader
+
+    from apps.common import readers
+    from . import definitions_constants as consts
+
+    # check network definition
+    r = BufferReader(definition)
+    expected_type_number = EthereumDefinitionType.NETWORK
+    # TODO: can't check equality of MsgDefObjs now, so we check the name
+    if expected_type.MESSAGE_NAME == EthereumTokenInfo.MESSAGE_NAME:
+        expected_type_number = EthereumDefinitionType.TOKEN
+
+    try:
+        # first check format version
+        if r.read_memoryview(len(consts.FORMAT_VERSION)) != consts.FORMAT_VERSION:
+            raise DataError("Invalid Ethereum definition")
+
+        # second check the type of the data
+        if r.get() != expected_type_number:
+            raise DataError("Definition type mismatch")
+
+        # third check data version
+        if readers.read_uint32_le(r) < consts.MIN_DATA_VERSION:
+            raise DataError("Definition is outdated")
+
+        # get payload
+        payload_length = readers.read_uint16_le(r)
```

Comment:
Nit: Not that it really matters, but the data format seems a bit chaotic, payload length is 2 bytes, proof length is 1 byte. We normally tend to use compact-size integers to length-prefix arrays as it is done in bitcoin.

Reply:
Oh wait, `proof_length` is not the length, more like `proof_entry_count` or `tree_depth`. But the same is true, e.g. the number of witness items is encoded as a compact-size integer too in bitcoin.

Reply:
I agree that it's kind of all over the place, but OTOH introducing compactsize seems like adding complexity just for the sake of making it look nice.
We originally considered using protobuf for the outer format as well, but in the end decided against it.